src/common/framework/gpu/device_pool.ts

```diff
@@ -114,56 +102,85 @@ class DescriptorToDeviceMap {
    *
    * Throws SkipTestCase if devices with this descriptor are unsupported.
    */
-  async getOrInsert(
-    key: GPUDeviceDescriptor,
-    create: () => Promise<DeviceHolder>
-  ): Promise<DeviceHolder> {
+  async getOrCreate(uncanonicalizedDescriptor: GPUDeviceDescriptor): Promise<DeviceHolder> {
+    const [descriptor, key] = canonicalizeDescriptor(uncanonicalizedDescriptor);
     // Never retry unsupported configurations.
-    for (const desc of this.unsupported) {
-      if (objectEquals(key, desc)) {
-        throw new SkipTestCase(`GPUDeviceDescriptor previously failed: ${JSON.stringify(key)}`);
-      }
+    if (this.unsupported.has(key)) {
+      throw new SkipTestCase(
+        `GPUDeviceDescriptor previously failed: ${JSON.stringify(descriptor)}`
+      );
```

Comment:
what about tests that intentionally fail creation? we don't want those tests to get skipped

Reply:
Tests that intentionally fail creation should call createDevice directly instead of going through this mechanism (which only exists to reuse existing valid devices)

Reply:
This mechanism is exposed as `t.selectDeviceOrSkipTestCase()` so the behavior (and the fact that it's not suited for such tests) should be clear.

Reply:
Ah okay, sounds good.